src/jdk.jlink/share/classes/jdk/tools/jlink/internal/TaskHelper.java

```diff
@@ -419,6 +420,77 @@ private PluginsConfiguration getPluginsConfig(Path output, Map<String, String> l
                 }
             }
 
+            // if we perform a run-time based link, add relevant exclude
+            // patterns, so as to match the packaged-modules-based link
+            if (!config.useModulePath()) {
+                Plugin systemModulesPlugin = null;
+                Plugin excludeResourcePlugin = null;
+                Plugin addOptionsPlugin = null;
+                for (Plugin p: pluginToMaps.keySet()) {
+                    if (p instanceof ExcludePlugin) {
+                        excludeResourcePlugin = p;
+                    }
+                    if (p instanceof SystemModulesPlugin) {
+                        systemModulesPlugin = p;
+                    }
+                    if (p instanceof AddOptionsPlugin) {
+                        addOptionsPlugin = p;
+                    }
+                }
+                // Certain system module classes get generated and SystemModulesMap replaced by the
+                // SystemModulesPlugin. Filter previously generated classes so the
+                // set of classes match the set of the packaged modules link.
+                String systemModulesPattern = "regex:/java\\.base/jdk/internal/module/SystemModules\\$.*\\.class";
+                // The default OpenJDK build generates JLI Species classes so they
+                // would be in the run-time image we derive from if we don't filter
+                // them. This filter corresponds to the set of classes part of the
+                // jmods.
+                String speciesPattern = "regex:/java\\.base/java/lang/invoke/BoundMethodHandle\\$Species_(?:D|DL|I|IL|LJ|LL).*\\.class";
+                // The AddResourcePlugin (used by AddOptionsPlugin)
+                String addOptionsGlob = "glob:" + AddOptionsPlugin.OPTS_FILE;
+                String saveJlinkOptsGlob = "glob:/jdk.jlink/" + JlinkTask.OPTIONS_RESOURCE;
+                String additionalPatterns = systemModulesPattern + "," +
```

Comment:
Suggest to have each plugin to define `Plugin::excludeResourcesPattern` to return a non-empty string if linking from the run-time image.  Construct this pattern using the API.